File: Leetcode/Python/5. Longest Palindromic Substring.py
class Solution(object):
    def longestPalindrome(self, s):
        if len(s) == 0:
            return ""
        start = 0
        end = 0
        def expandAroundCenter(left, right):
            while left >= 0 and right < len(s) and s[left] == s[right]:
                left -= 1
                right += 1
            return left+1, right-1
        
        for i in range(len(s)):
            left1, right1 = expandAroundCenter(i, i)
            left2, right2 = expandAroundCenter(i, i+1)
            if right1 - left1 > end - start:
                start, end = left1, right1
            if right2 - left2 > end - start:
                start, end = left2, right2
        return s[start:end+1]

if __name__ == '__main__':
    s = Solution()
    print(s.longestPalindrome("babad"))
    print(s.longestPalindrome("cbbd"))
    print(s.longestPalindrome("cbbbaaaa"))

File: Leetcode/PHP/5. Longest Palindromic Substring.php
<?php
class Solution {
    function longestPalindrome($s) {
        if (strlen($s) == 0) return "";

        $start = 0;
        $end = 0;

        for ($i = 0; $i < strlen($s); $i++) {
            list($left1, $right1) = $this->expandAroundCenter($s, $i, $i);
            list($left2, $right2) = $this->expandAroundCenter($s, $i, $i + 1);

            if ($right1 - $left1 > $end - $start) {
                $start = $left1;
                $end = $right1;
            }
            if ($right2 - $left2 > $end - $start) {
                $start = $left2;
                $end = $right2;
            }
        }

        return substr($s, $start, $end - $start + 1);
    }

    private function expandAroundCenter($s, $left, $right) {
        while ($left >= 0 && $right < strlen($s) && $s[$left] == $s[$right]) {
            $left--;
            $right++;
        }
        return [$left + 1, $right - 1];
    }
}


$s = new Solution();
echo $s->longestPalindrome("babad") . "\n";
echo $s->longestPalindrome("cbbd") . "\n";
echo $s->longestPalindrome("cbbbaaaa") . "\n";
?>
